sdk/keyvault/azure-security-keyvault-administration/src/main/java/com/azure/security/keyvault/administration/KeyVaultAccessControlAsyncClient.java

```diff
@@ -185,32 +211,330 @@ HttpPipeline getHttpPipeline() {
+                            .map(allowedDataAction -> DataAction.fromString(allowedDataAction.toString()))
+                            .collect(Collectors.toList()))
+                        .setNotDataActions(keyVaultPermission.getNotDataActions().stream()
+                            .map(notDataAction -> DataAction.fromString(notDataAction.toString()))
+                            .collect(Collectors.toList())))
+                    .collect(Collectors.toList());
+            }
+
+            RoleDefinitionProperties roleDefinitionProperties =
+                new RoleDefinitionProperties()
+                    .setRoleName(options.getRoleDefinitionName())
+                    .setRoleType(RoleType.CUSTOM_ROLE)
+                    .setAssignableScopes(assignableScopes)
+                    .setDescription(options.getDescription())
+                    .setPermissions(permissions);
+            RoleDefinitionCreateParameters parameters =
+                new RoleDefinitionCreateParameters()
+                    .setProperties(roleDefinitionProperties);
+
+            return clientImpl.getRoleDefinitions()
+                .createOrUpdateWithResponseAsync(vaultUrl, options.getRoleScope().toString(),
+                    options.getRoleDefinitionName(), parameters,
+                    context.addData(AZ_TRACING_NAMESPACE_KEY, KEYVAULT_TRACING_NAMESPACE_VALUE))
+                .doOnRequest(ignored ->
+                    logger.verbose("Creating role assignment - {}", options.getRoleDefinitionName()))
+                .doOnSuccess(response -> logger.verbose("Created role assignment - {}", response.getValue().getName()))
+                .doOnError(error ->
+                    logger.warning("Failed to create role assignment - {}", options.getRoleDefinitionName(), error))
+                .onErrorMap(KeyVaultAdministrationUtils::mapThrowableToKeyVaultAdministrationException)
+                .map(KeyVaultAccessControlAsyncClient::transformRoleDefinitionResponse);
+        } catch (RuntimeException e) {
+            return monoError(logger, e);
+        }
+    }
+
+    /**
+     * Gets a {@link KeyVaultRoleDefinition}.
+     *
+     * @param roleScope The {@link KeyVaultRoleScope role scope} of the {@link KeyVaultRoleDefinition}.
+     * @param roleDefinitionName The name used of the {@link KeyVaultRoleDefinition}.
+     *
+     * @return A {@link Mono} containing the {@link KeyVaultRoleDefinition}.
+     *
+     * @throws KeyVaultAdministrationException If a {@link KeyVaultRoleDefinition role definition} with the given name cannot
+     * be found or if the given {@code roleScope} is invalid.
+     * @throws NullPointerException If the {@link KeyVaultRoleScope role scope} or {@link String roleDefinitionName} are
+     * {@code null}.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Mono<KeyVaultRoleDefinition> getRoleDefinition(KeyVaultRoleScope roleScope, String roleDefinitionName) {
```

Comment:
Do you have a GUID type you could use for any `roleDefinitionName`? To note, though, any `roleAssignmentId` should be a string. I forget the reason off hand, but there was one. Maybe @christothes remembers.

Reply:
In a previous discussion we decided against using the `UUID` type and let the service do the validation, at least for Java.